en/_posts/2024-05-23-understanding-github-artifact-attestations.md

```diff
@@ -0,0 +1,224 @@
+artifact file.
+
+```
+- name: Attest Build Provenance
+  uses: actions/attest-build-provenance@897ed5eab6ed058a474202017ada7f40bfa52940 # v1.0.0
+  with:
+    subject-path: "bin/my-artifact.tar.gz"
+```
+
+Easy-to-use UX is really important for security because it increases adoption.
+This can’t be understated and is an often overlooked aspect of security
+software.
+
+## Why SLSA Build L2?
+
+But if this is a fairly secure solution, why is this SLSA Build L2 and not L3?
+The main reason is that, by itself, the `attest-build-provenance` action
+doesn’t meet this requirement for [SLSA Build
+L3](https://slsa.dev/spec/v1.0/levels#build-l3-hardened-builds):
+
+prevent secret material used to sign the provenance from being accessible to the user-defined build steps.
+
+Because GitHub Actions are run in the same job VM with other build steps there
+is a chance that the user-defined build steps could access the secret material,
+namely the certificate’s private key provided by Fulcio, and use it for
+nefarious purposes. Normally an attacker, by compromising the build, might be
+able to use this key material to make up their own provenance for a malicious
+artifact and sign it with the key.
+
+However, as we’ll see, this kind of attack is mostly mitigated by using Sigstore.
+
+## SLSA Build L2+?
+
+By using Sigstore’s Fulcio the certificate used to sign the provenance contains
+much of the provenance information in its OID claims. These claims are signed
+by part of Fulcio instance’s certificate chain and thus cannot be modified by
+the user-defined build steps unless the Sigstore instance or GitHub’s OIDC
+provider are compromised.
+
+So while the SLSA provenance itself might be modified the certificate OID
+claims cannot. So GitHub can check the OID claims against the expected values
+in order to verify them even though the user-defined build steps had access to
+the signing key. So this is why verification doesn’t rely on the provenance
+itself and instead relies on the certificate’s OID claims for verification.
+
+Some folks have colloquially referred to this combination of Sigstore and SLSA
+Build L2 as SLSA Build L2+ since it provides some of the benefits of SLSA Build
+L3 without actually fulfilling all of the requirements of L3.
+
+## Limits of SLSA Build L2+
+
+However, this comes with a caveat. This means that if the user-defined build
+steps could have access to the signing keys then **only** the information in
+the certificate’s OID claims are trustworthy. No information that isn’t
+included in these claims can be included in the provenance because it can’t be
+verified later.
+
+Crucially this includes values like the GitHub Actions workflow
+[`inputs`](https://docs.github.com/en/actions/learn-github-actions/contexts#inputs-context)
+and
+[`vars`](https://docs.github.com/en/actions/learn-github-actions/contexts#vars-context)
+contexts. These include the inputs into a build and could potentially be used
```

Comment:
I this this also means: If we wanted to do offline verification, without using github's cli to re-download the provenance, then not even the artifact hash can be trusted, due to the replay attack potential you mentioned in the previous section.

Reply:
In reality, the hash is whatever the build gives you (or at least it's the hash of whatever file the build gives you) so it's not an especially trustworthy value (since provenance says nothing about whether builds themselves were compromised).

If the build is compromised it could just give the SLSA builder a malicious file to sign and have it sign it without needing the key material. I guess the only difference is that an attacker would need an actual build to run in this case. If they had the key material they could sign an arbitrary number of files from anywhere without actually needing CI or a build to run (though they'd only have about 10 min from when the key was minted to do that).

Reply:
That's what I was getting at, like we talked about before. But to expound on the story: The attacker could surreptitiously steal the keys without modifying the original build, so as to not alert the maintainers. Then the attacker could sign and distribute their own artifacts on a fake package mirror. For consumers that have provenance verification already automated, they might not be susceptible to using a malicious mirror. But it could happen - https://checkmarx.com/blog/over-170k-users-affected-by-attack-using-fake-python-infrastructure/

Reply:
Do you think I should update this section somehow to convey this better?

Reply:
I think so. Similar to the way you put it about an attacker distributing signed files and attestations without leaving any trace.

Reply:
Updated the post to incorporate this. LMK what you think.